sd-validation/schema.js

```diff
@@ -0,0 +1,148 @@
+/**
+ * @license Copyright 2018 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const walkObject = require('./helpers/walkObject');
+// @ts-ignore
+const schemaStructure = new Map(require('./assets/schema_google'));
+const TYPE_KEYWORD = '@type';
+const SCHEMA_ORG_URL = 'http://schema.org/';
+
+/**
+ * @param {string} uri
+ * @returns {string}
+ */
+function cleanName(uri) {
+  return uri.replace(SCHEMA_ORG_URL, '');
+}
+
+/**
+ * @param {string} type
+ * @returns {{props: Array<string>, required: ?Array<string>, recommended: ?Array<string>}}
+ */
+function getTypeSettingsForType(type) {
+  const typeSettings = schemaStructure.get(type);
+
+  return typeSettings;
+}
+
+/**
+ * @param {string} type
+ * @returns {boolean}
+ */
+function isKnownType(type) {
+  return schemaStructure.has(type);
+}
+
+/**
+ * Validates keys of given object based on it's type(s). Returns an array of error messages.
+ *
+ * @param {String|Array<String>} typeOrTypes
+ * @param {Array<String>} keys
+ * @returns {Array<String>}
+ */
+function validateObjectKeys(typeOrTypes, keys) {
+  /** @type {Array<string>} */
+  const errors = [];
+  /** @type {Array<string>} */
+  const safelist = [];
+  /** @type {Array<string>} */
+  const required = [];
+  /** @type {Array<string>} */
+  const recommended = [];
+
+  let types = [];
+
+  if (typeof typeOrTypes === 'string') {
+    types.push(typeOrTypes);
+  } else if (Array.isArray(typeOrTypes)) {
+    types = typeOrTypes;
+  } else {
+    return ['Unknown value type'];
+  }
+
+  const unknownTypes = types.filter(t => !isKnownType(t));
+
+  unknownTypes
+    .forEach(type => {
+      if (typeof type !== 'string' || type.indexOf(SCHEMA_ORG_URL) === 0) {
+        errors.push(`Unrecognized schema.org type ${type}`);
+      }
+    });
+
+  if (unknownTypes && unknownTypes.length) {
+    return errors;
+  }
+
+  types.forEach(type => {
+    const typeSettings = getTypeSettingsForType(type);
+
+    if (typeSettings.props) {
+      typeSettings.props.forEach(key => safelist.push(key));
+    }
+
+    if (typeSettings.required) {
+      typeSettings.required.forEach(key => required.push(key));
+    }
+
+    if (typeSettings.recommended) {
+      typeSettings.recommended.forEach(key => recommended.push(key));
+    }
+  });
+
+  const cleanKeys = keys
+    // skip JSON-LD keywords
+    .filter(key => key.indexOf('@') !== 0)
```

Comment:
in jsonld.js you have that safelist of keywords. why not use that explicit list?

Reply:
ok cool. then we can just augment the comment 

    // Skip JSON-LD keywords (including invalid ones as they were already flagged in the json-ld validator)

Reply:
Fixed by Konrad.